### DISCUSSION TITLE
Is there any way to debug TypeScript Serverless Functions?

### DISCUSSION THREAD

**Discussion by lucas-varela** [vercel/vercel#4478] [Help] [answered]
I'm relatively new to Vercel CLI (and TypeScript, for that matter) and I'm currently writing an API using Serverless Functions. Everything is working fine, but I'm not sure how to debug my code when using `vercel dev`.

As far as I understood, I need to point my IDE to the compiled JS files to debug them, but I'm not sure where these are saved by the CLI. Where can I find these built files? Or am I supposed to debug this kind of code using other methods??

Thanks in advance :)

**Comment by ctrlplusb** [comment]
What do you mean "debug"?

I have written a util that helps me unit test my functions in isolation. Would work with TS too. May be helpful in your case?

https://github.com/ctrlplusb/vercel-node-server

**Reply by lucas-varela** [reply]
By "debug", I meant pausing the server process with breakpoints. I've written what I've found out below. Anyway, I didn't test your package because it would be cumbersome to wrap my functions on a separate file every time I wanted to debug something. That being said, however, I've read your code and I'm pretty sure it would work as a workaround if anyone is having an awful time trying to debug something without a real debugger when using `vercel dev`. So thank you very much for taking the time to showing an alternative for the time being. And being able to unit test functions is awesome.

**Reply by ctrlplusb** [reply]
Hey @lucas-varela 

I have been working on an [alternative development CLI](https://github.com/ctrlplusb/vercel-node-dev), which supports attaching a debugger session. It doesn't support Next.js, and only supports Node.js runtime. If that matches your project configuration then I would appreciate if you could give it a try and feedback to me. 👍

**Comment by lucas-varela** [comment, marked as answer]
So, I found out that this is not a TypeScript specific bug. In short, `vercel dev` currently does not support Node debugging. There's a closed issue about it here: #2864.

However, [as stated here](https://github.com/zeit/now/issues/3849#issuecomment-620157834), the Vercel team is already working on rewriting the `dev` command, and debug support will be added.

For people with problems with the debugger not pausing on breakpoints: hang in there and keep posted about the CLI updates.

**Reply by angryalbatross** [reply]
Thanks Lucas.  I have spent days trying to figure this out between google and stackoverflow.  Should have come here first!

**Comment by holylander** [comment]
Hi there! 
I believe this example may be useful for oders.
In order to debug your `api` vercel serverless functions with the help of `vscode`:

1. Init the vercel dev environment this way: `NODE_OPTIONS='--inspect' vercel dev`
2. Init the debugging session. Use this configuration:
```
        {
            "type": "node",
            "request": "attach",
            "name": "Attach vercel dev",
            "skipFiles": [
                "<node_internals>/**"
            ],
            "port": 9229,
            "sourceMaps": true,
            "cwd": "${workspaceRoot}",
        },
```
With this setup, you should be able to query you serverless local endpoint, and stop its processing when a breakpoint is found ;)

**Reply by danechitoaie** [reply]
Does this work with TypeScript functions as well?

All I get is endless messages like the following:
```
╰─ λ NODE_OPTIONS='--inspect' vercel dev
Debugger listening on ws://127.0.0.1:9229/efc80254-5147-4db8-bb90-0de3f41c5985
For help, see: https://nodejs.org/en/docs/inspector
Vercel CLI 23.1.2 dev (beta) — https://vercel.com/feedback
> Ready! Available at http://localhost:3000
Debugger attached.
Starting inspector on 127.0.0.1:9229 failed: address already in use
Starting inspector on 127.0.0.1:9229 failed: address already in use
Starting inspector on 127.0.0.1:9229 failed: address already in use
Starting inspector on 127.0.0.1:9229 failed: address already in use
Starting inspector on 127.0.0.1:9229 failed: address already in use
Starting inspector on 127.0.0.1:9229 failed: address already in use
Starting inspector on 127.0.0.1:9229 failed: address already in use
Starting inspector on 127.0.0.1:9229 failed: address already in use
Starting inspector on 127.0.0.1:9229 failed: address already in use
Starting inspector on 127.0.0.1:9229 failed: address already in use
Starting inspector on 127.0.0.1:9229 failed: address already in use
Starting inspector on 127.0.0.1:9229 failed: address already in use
Starting inspector on 127.0.0.1:9229 failed: address already in use
Starting inspector on 127.0.0.1:9229 failed: address already in use
Starting inspector on 127.0.0.1:9229 failed: address already in use
Starting inspector on 127.0.0.1:9229 failed: address already in use
Starting inspector on 127.0.0.1:9229 failed: address already in use
Starting inspector on 127.0.0.1:9229 failed: address already in use
Starting inspector on 127.0.0.1:9229 failed: address already in use
Starting inspector on 127.0.0.1:9229 failed: address already in use
```

And the breakpoints are ignored.

**Reply by holylander** [reply]
ur port 9229 is in use by some other process. Each port can only be in use by a single process

**Reply by danechitoaie** [reply]
it's used by vercel or by vercel spawning processes as the first time it starts ok. I only start getting the message after I start the debugger and connect to the process and make the first request.

**Reply by danechitoaie** [reply]
```
╰─ λ lsof | grep 9229                                                                                                                                      ↵ 130
node      10259 daniels   26u     IPv4 0x3f743dd9b57c90ef        0t0                 TCP localhost:9229 (LISTEN)
node      10259 daniels   53u     IPv4 0x3f743dd9c62840ef        0t0                 TCP localhost:9229->localhost:53091 (ESTABLISHED)
Code\x20H 10275 daniels   61u     IPv4 0x3f743dd9b5f860ef        0t0                 TCP localhost:53091->localhost:9229 (ESTABLISHED)
```

**Reply by holylander** [reply]
dont know how is ur debug config, but these are the steps:
run `NODE_ENV=test NODE_OPTIONS='--inspect' vercel dev`
then I start a debugger profile with this config:
```
        {
            "name": "Debug 'vercel dev' process - ENV:test",
            "type": "node",
            "request": "attach",
            "skipFiles": [
                "<node_internals>/**"
            ],
            "env": {
                "NODE_ENV": "test"
            },
            "port": 9229,
            "sourceMaps": true,
            "cwd": "${workspaceRoot}",
        },
```
Hope it helps you ;)

**Comment by huantbui** [comment]
I got vscode to debug with `vercel dev`.

In your `package.json`, create any script name (eg: "start:local": "vercel dev").  
Click on a break point that you want your code to stop at a specific API route.
Hover over the "start:local" and run as: "Debug Script". Call that API route and it should hit that endpoint with your data.

![vscode-vercel-dev](https://user-images.githubusercontent.com/14843167/179890094-91d41d0f-c7d9-4760-b420-59ed7ee7a86f.gif)
